src/grain/createGrainDistribution.js

```diff
@@ -0,0 +1,269 @@
+// @flow
+
+/**
+ * In SourceCred, projects regularly distribute Grain to contributors based on
+ * their Cred scores. This is called a "Distribution".
+ *
+ * This module contains the logic for calculating distribution amounts for
+ * contributors. A distribution contains "receipts" showing how much Grain each
+ * contributor will receive, the timestamp of the distribution in question, and the
+ * "strategy" used to distribute grain.
+ *
+ * Currently we support two strategies:
+ * - IMMEDIATE, which distributes a fixed budget of grain based on the cred scores
+ * in the most recent completed time interval
+ * - BALANCED, which distributes a fixed budget of grain based on cred scores
+ * across
+ * all time, prioritizing paying people who were under-paid historically (i.e.
+ * their lifetime earnings are lower than we would expect given their current
+ * cred score)
+ *
+ * In all cases, the timestamp for the distribution is used to determine which cred
+ * scores are in scope. For example, if you create a immediate distribution with a
+ * timestamp in the past, it will reward people with cred in the past time
+ * period, not the current one.
+ */
+import {sum} from "d3-array";
+import {mapToArray} from "../util/map";
+import {type NodeAddressT} from "../core/graph";
+import {
+  type Grain,
+  multiplyFloat,
+  ZERO,
+  toFloatRatio,
+  format,
+  DECIMAL_PRECISION,
+} from "./grain";
+
+export const GRAIN_DISTRIBUTION_VERSION_1 = 1;
+
+export type DistributionStrategy = ImmediateV1 | BalancedV1;
+
+export type ImmediateV1 = {|
+  +type: "IMMEDIATE",
+  +version: number,
+  +budget: Grain,
+|};
+
+export type BalancedV1 = {|
+  +type: "BALANCED",
+  +version: number,
+  +budget: Grain,
+|};
+
+export type GrainReceipt = {|
+  +address: NodeAddressT,
+  +amount: Grain,
+|};
+
+export type GrainDistributionV1 = {|
+  +type: "GRAIN_DISTRIBUTION",
```

Comment:
I believe @Beanow was suggesting that we use past-tense names, i.e. "GRAIN_DISTRIBUTED".

However, I am gonna churn things up a bit, and say: Maybe we should de-couple this type from the ledger event type?

The only reason we have a `TYPE` name here is so that we can distinguish it from other ledger events. However, it's a bit weird that some ledger events are not owned directly by the ledger, but by associated modules. 

Also, we have a timestamp argument here, which is necessary for the ledger, but is not obviously necessary for the distribution. Especially pursuant to [this comment thread](https://github.com/sourcecred/sourcecred/pull/1743#discussion_r409058194), we might want to move all explicitly time-based calculation out of the distribution module. And we already want to make a change so that we have multiple sub-distributions inside a single distribution, per #1760.

So, I think we can kill a few birds with one stone here: (though why would we want to kill birds? 🤔🐦)
- remove the type and timestamp fields from the GrainDistribution (maybe rename to something like GrainAllocation)
- in the ledger, we define a GrainDistributed type which has a timestamp, type, and an array of GrainAllocations

Now we simplify this module, and we wind up with the outcome that I believe @Beanow wanted, which is that we can immutably "seal" a few allocations together.

The method which produces the GrainDistribution from a set of strategies can own the logic for making sure they are applied "in order" in a consistent way (i.e. keeping local state about the updated earnings).

what do you think @hammadj @Beanow

Reply:
I would say, that is a good suggestion for this PR.

But to unblock the grain ledger in general, I think it's best to take a step back and make sure we agree on architecture and requirements as that's what most of the discussion is about. When we've settled on those, it should be much easier to implement and review things.

Reply:
I agree with the suggestion, I assume we also want to rename `DistributionStrategy` to `AllocationStrategy` and `createGrainDistribution` to `createGrainAllocation`? The reason I didnt change it to "GRAIN_DISTRIBUTED" was because of the naming of the rest of the stuff (`createGrainDistributedEvent` felt weird since the event stuff wasn't fully fleshed out, and `distributeGrain` didnt feel right either since the function was just calculating the allocation, not submitting it to the ledger or anything). Renaming to allocation and having a "distribution event" hold an array of allocation makes more sense to me than having an array of strategies / receipts within a single allocation.